bluemira/builders/tf_coils.py

```diff
@@ -140,17 +143,397 @@ def _make_single_circuit(self, wire):
+        if grad.size > 0:
+            grad[:] = approx_derivative(
+                func,
+                vector,
+                f0=constraint,
+                args=(parameterisation, solver, points, TF_ripple_limit),
+                **ad_args,
+            )
+
+        bluemira_debug_flush(f"Max ripple: {max(constraint+TF_ripple_limit)}")
+        return constraint
+
+    @staticmethod
+    def calculate_ripple(vector, parameterisation, solver, points, TF_ripple_limit):
+        """
+        Calculate ripple constraint
+
+        Parameters
+        ----------
+        vector: np.ndarray
+            Variable vector
+        parameterisation: GeometryParameterisation
+            Geometry parameterisation
+        solver: ParameterisedHelmholtzSolver
+            TF ripple solver
+        points: Coordinates
+            Coordinates at which to calculate the ripple
+        TF_ripple_limit: float
+            Maximum allowable TF ripple
+
+        Returns
+        -------
+        c_values: np.ndarray
+            Ripple constraint values
+        """
+        parameterisation.variables.set_values_from_norm(vector)
+        wire = parameterisation.create_shape()
+        solver.update_cage(wire)
+        ripple = solver.ripple(*points)
+        return ripple - TF_ripple_limit
+
+
+class EquispacedSelector(RipplePointSelector):
+    """
+    Equispaced ripple points along a wire for a given number of points.
+
+    Parameters
+    ----------
+    n_rip_points:
+        Number of points along the wire constrain the ripple
+    x_frac:
+        If specified, the fraction of radius above which the points will
+        be selected.
+        If unspecified, the points will be selected on the full wire
+    """
+
+    def __init__(self, n_rip_points: int, x_frac: Optional[float] = None):
+        self.n_rip_points = n_rip_points
+        self.x_frac = x_frac
+
+    def set_wire(self, wire):
+        """
+        Set the wire along which the points will be selected
+
+        Parameters
+        ----------
+        wire
+            Wire along which the points will be selected
+        """
+        super().set_wire(wire)
+        if self.x_frac is not None and not np.isclose(self.x_frac, 0.0):
+            self.x_frac = np.clip(self.x_frac, 0.005, 0.995)
+            bb = wire.bounding_box
+
+            x_min = bb.x_min + self.x_frac * (bb.x_max - bb.x_min)
+
+            z_min, z_max = bb.z_min - 10, bb.z_max + 10
+            cut_face = BluemiraFace(
+                make_polygon(
+                    {
+                        "x": [0, x_min, x_min, 0],
+                        "y": 0,
+                        "z": [z_min, z_min, z_max, z_max],
+                    },
+                    closed=True,
+                )
+            )
+            wire = boolean_cut(wire, cut_face)[0]
+        self.points = wire.discretize(byedges=True, ndiscr=self.n_rip_points)
```

Comment:
Is `self.points` not consistent with `self.wire` as you possibly modify the original wire to make the points?

Reply:
`self.wire` should be set with `set_wire`. The wire is not modified by any of the classes, just used as a basis from which to get the points. I will make it a private attribute, to discourage direct attribution.